cmd/reva/app-tokens-create.go

```diff
@@ -0,0 +1,257 @@
+	"time"
+
+	authapp "github.com/cs3org/go-cs3apis/cs3/auth/applications/v1beta1"
+	authpb "github.com/cs3org/go-cs3apis/cs3/auth/provider/v1beta1"
+	gateway "github.com/cs3org/go-cs3apis/cs3/gateway/v1beta1"
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	share "github.com/cs3org/go-cs3apis/cs3/sharing/link/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	types "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	"github.com/cs3org/reva/pkg/auth/scope"
+	"github.com/cs3org/reva/pkg/errtypes"
+)
+
+type appTokenCreateOpts struct {
+	Expiration string
+	Label      string
+	Path       stringSlice
+	Share      stringSlice
+	Unlimited  bool
+}
+
+type stringSlice []string
+
+func (ss *stringSlice) Set(value string) error {
+	*ss = append(*ss, value)
+	return nil
+}
+
+func (ss *stringSlice) String() string {
+	return strings.Join([]string(*ss), ",")
+}
+
+var createOpts *appTokenCreateOpts = &appTokenCreateOpts{}
+
+const layoutTime = "2006-01-02"
+
+func appTokensCreateCommand() *command {
+	cmd := newCommand("token-create")
+	cmd.Description = func() string { return "create a new application tokens" }
+	cmd.Usage = func() string { return "Usage: token-create" }
+
+	cmd.StringVar(&createOpts.Label, "label", "", "set a label")
+	cmd.StringVar(&createOpts.Expiration, "expiration", "", "set expiration time (format <yyyy-mm-dd>)")
+	// TODO(gmgigi96): add support for multiple paths and shares for the same token
+	cmd.Var(&createOpts.Path, "path", "create a token for a file (format path:[r|w]). It is possible specify multiple times this flag")
+	cmd.Var(&createOpts.Share, "share", "create a token for a share (format shareid:[r|w]). It is possible specify multiple times this flags")
+	cmd.BoolVar(&createOpts.Unlimited, "all", false, "create a token with an unlimited scope")
+
+	cmd.ResetFlags = func() {
+		s := reflect.ValueOf(createOpts).Elem()
+		s.Set(reflect.Zero(s.Type()))
+	}
+
+	cmd.Action = func(w ...io.Writer) error {
+
+		err := checkOpts(createOpts)
+		if err != nil {
+			return err
+		}
+
+		client, err := getClient()
+		if err != nil {
+			return err
+		}
+
+		ctx := getAuthContext()
+
+		scope, err := getScope(ctx, client, createOpts)
+		if err != nil {
+			return err
+		}
+
+		// parse eventually expiration time
+		var expiration *types.Timestamp
+		if createOpts.Expiration != "" {
+			exp, err := time.Parse(layoutTime, createOpts.Expiration)
+			if err != nil {
+				return err
+			}
+			expiration = &types.Timestamp{
+				Seconds: uint64(exp.Unix()),
+			}
+		}
+
+		generateAppPasswordResponse, err := client.GenerateAppPassword(ctx, &authapp.GenerateAppPasswordRequest{
+			Expiration: expiration,
+			Label:      createOpts.Label,
+			TokenScope: scope,
+		})
+
+		if err != nil {
+			return err
+		}
+		if generateAppPasswordResponse.Status.Code != rpc.Code_CODE_OK {
+			return formatError(generateAppPasswordResponse.Status)
+		}
+
+		pw := generateAppPasswordResponse.AppPassword
+
+		if len(w) == 0 {
```

Comment:
Remove this check and just print the output. This is needed only for commands used in auto-complete